tests/stub/driver_parameters/test_bookmark_manager.py

```diff
@@ -0,0 +1,598 @@
+import json
+import re
+
+from nutkit.frontend import (
+    Driver,
+    Neo4jBookmarkManagerConfig,
+)
+import nutkit.protocol as types
+from tests.shared import TestkitTestCase
+from tests.stub.shared import StubServer
+
+
+class TestNeo4jBookmarkManager(TestkitTestCase):
+    required_features = (
+        types.Feature.BOLT_5_0,
+        types.Feature.API_BOOKMARK_MANAGER,
+    )
+
+    def setUp(self):
+        super().setUp()
+        self._server = StubServer(9010)
+        self._router = StubServer(9000)
+        self._driver = None
+        self._session = None
+        self._sessions = []
+
+    def tearDown(self):
+        self._server.reset()
+        self._router.reset()
+
+        for s in self._sessions:
+            s.close()
+
+        if self._session:
+            self._session.close()
+
+        if self._driver:
+            self._driver.close()
+
+        return super().tearDown()
+
+    def test_should_keep_track_of_session_run(self):
+        self._start_server(self._router, "router_with_db_name.script")
+        self._start_server(self._server, "session_run_chaining.script")
+
+        uri = "neo4j://%s" % self._router.address
+        auth = types.AuthorizationToken("basic", principal="neo4j",
+                                        credentials="pass")
+        self._driver = Driver(
+            self._backend,
+            uri, auth,
+            bookmark_manager_config=Neo4jBookmarkManagerConfig()
+        )
```

Comment:
These lines could probably be refactored into a method

```python
    def _connect_driver(self, bookmark_config=None):
        if bookmark_config is None:
            bookmark_config = Neo4jBookmarkManagerConfig()
        uri = "neo4j://%s" % self._router.address
        auth = types.AuthorizationToken("basic", principal="neo4j",
                                        credentials="pass")
        self._driver = Driver(self._backend, uri, auth,
                              bookmark_manager_config=bookmark_config)
```

or something. It's being duplicated across many tests.